src/lib/poptrie_lookup.dasl

```diff
@@ -0,0 +1,146 @@
+      dasm.dump(mcode, size)
+   end
+
+   local prototype
+   if keysize <= 64 then
+      prototype = ffi.typeof(
+         "$ (*) ($ *, $ *, uint64_t, $ *)",
+         Poptrie.leaf_t, Poptrie.leaf_t, Poptrie.node_t, Poptrie.base_t
+      )
+   else error("NYI") end
+
+   return ffi.cast(prototype, mcode)
+end
+
+|.define leaves,  rdi -- pointer to leaves array
+|.define nodes,   rsi -- pointer to nodes array
+|.define key,     rdx -- key to look up
+|.define dmap,    rcx -- pointer to directmap
+|.define index,   r8d -- index into node array
+|.define node,    r8  -- pointer into node array 
+|.define offset,  r9  -- offset into key
+|.define v,       r10 -- k or s bits extracted from key
+|.define vec,     r11 -- 64-bit vector or leafvec
+
+-- lookup(leaf_t *leaves, node_t *nodes, key) -> leaf_t
+function lookup (Dst, Poptrie, keysize)
+   if Poptrie.direct_pointing then
+      -- v = extract(key, 0, Poptrie.s)
+      local direct_mask = bit.lshift(1ULL, Poptrie.s) - 1
+      -- v = band(key, direct_mask)
+      | mov v, key
+      | and v, direct_mask
+      -- index = dmap[v]
+      | mov index, dword [dmap+v*4]
+      -- eax = band(index, leaf_tag - 1) (tag inverted)
+      | mov eax, index
+      -- is leaf_tag set? (unsets bit)
+      | btr eax, 31
+      | jnc >1 -- leaf_tag not set, index is a node
+      | ret
+      -- node, offset = nodes[index], s
+      |1:
+      | imul index, 24 -- multiply by node size
+      | lea node, [nodes+index]
+      -- offset = s
+      | mov offset, 18
+   else
+      -- index, node, offset = 0, nodes[index], 0
+      | xor index, index
+      | lea node, [nodes+0] -- nodes[0]
+      | xor offset, offset
+   end
+   -- while band(vec, lshift(1ULL, v)) ~= 0
+   |2:
+   -- v = extract(key, offset, k=6)
+   if keysize <= 64 then
+      -- v = rshift(key, offset)
+      | mov v, key
+      | mov rcx, offset
+      | shr v, cl
+      -- v = band(v, lshift(1, k=6) - 1)
+      | and v, 0x3F
+   else error("NYI") end
+   -- vec = nodes[index].vector
+   | mov vec, qword [node+8]
+   -- is bit v set in vec?
+   | bt vec, v
+   | jnc >3 -- reached leaf, exit loop
+   -- rax = lshift(2ULL, v) - 1
+   | mov rax, 2
+   | mov rcx, v
+   | shl rax, cl
+   | sub rax, 1
+   -- rax = popcnt(band(vec, rax))
+   | and rax, vec
+   | popcnt rax, rax
+   -- index = base + bc - 1
+   | mov index, dword [node+20] -- nodes[index].base1
+   | sub index, 1
+   | add index, eax
+   -- node = nodes[index]
+   | imul index, 24 -- multiply by node size
+   | lea node, [nodes+index]
+   -- offset = offset + k
+   | add offset, 6
+   | jmp <2 -- loop
+   -- end while
+   |3:
+   -- rax = lshift(2ULL, v) - 1
+   | mov rax, 2
+   | mov rcx, v
+   | shl rax, cl
+   | sub rax, 1
+   if Poptrie.leaf_compression then
+      -- vec = nodes[index].leafvec
+      | mov vec, qword [node+0]
+   else error("NYI") end
+   -- rax = popcnt(band(vec, rax)) - 1
+   | and rax, vec
+   | popcnt rax, rax
+   -- return leaves[base + bc - 1]
```

Comment:
The `mov` and the `add` here could be fused, as in:
```
   | add eax, dword [node+16] -- nodes[index].base0
   | movzx eax, word [leaves+eax*2-2] -- leaves[index]
```
(No improvement in unfused-domain uops, but does save one fused-domain uop per function invocation, if that is even measurable against noise)